File: flowforge-common/src/main/java/com/flowforge/common/model/PlanLimits.java
package com.flowforge.common.model;

import java.util.Objects;

/**
 * Central plan limits configuration — single source of truth for all plan-based
 * resource limits. Used by every backend service for enforcement.
 *
 * A value of -1 means "unlimited".
 */
public class PlanLimits {

    private final int maxWorkflows;
    private final int maxModels;
    private final int maxExecutionsPerMonth;
    private final int maxTeamMembers;
    private final int reqPerMinute;
    private final int burstCapacity;
    private final int maxWebhooksPerDay;
    private final int priceMonthly; // cents (0 = free, -1 = custom/contact sales)

    private PlanLimits(int maxWorkflows, int maxModels, int maxExecutionsPerMonth,
                       int maxTeamMembers, int reqPerMinute, int burstCapacity,
                       int maxWebhooksPerDay, int priceMonthly) {
        this.maxWorkflows = maxWorkflows;
        this.maxModels = maxModels;
        this.maxExecutionsPerMonth = maxExecutionsPerMonth;
        this.maxTeamMembers = maxTeamMembers;
        this.reqPerMinute = reqPerMinute;
        this.burstCapacity = burstCapacity;
        this.maxWebhooksPerDay = maxWebhooksPerDay;
        this.priceMonthly = priceMonthly;
    }

    /**
     * Returns the plan limits for the given plan tier.
     */
    public static PlanLimits forPlan(Client.Plan plan) {
        return switch (plan) {
            case FREE -> new PlanLimits(
                    3,       // workflows
                    5,       // models
                    1_000,   // executions/month
                    2,       // team members
                    10,      // req/min
                    20,      // burst capacity
                    100,     // webhooks/day
                    0        // free
            );
            case PRO -> new PlanLimits(
                    25,      // workflows
                    50,      // models
                    100_000, // executions/month
                    10,      // team members
                    60,      // req/min
                    100,     // burst capacity
                    10_000,  // webhooks/day
                    4_900    // $49/mo
            );
            case ENTERPRISE -> new PlanLimits(
                    -1,      // unlimited
                    -1,
                    -1,
                    -1,
                    600,     // req/min
                    1_000,   // burst capacity
                    -1,
                    -1       // custom pricing
            );
        };
    }

    /**
     * Checks if the current count exceeds or meets this limit.
     * Returns false (not exceeded) when the limit is -1 (unlimited).
     */
    public static boolean isExceeded(int limit, long currentCount) {
        return limit != -1 && currentCount >= limit;
    }

    // ── Getters ──────────────────────────────────────────────────────────

    public int getMaxWorkflows() { return maxWorkflows; }
    public int getMaxModels() { return maxModels; }
    public int getMaxExecutionsPerMonth() { return maxExecutionsPerMonth; }
    public int getMaxTeamMembers() { return maxTeamMembers; }
    public int getReqPerMinute() { return reqPerMinute; }
    public int getBurstCapacity() { return burstCapacity; }
    public int getMaxWebhooksPerDay() { return maxWebhooksPerDay; }
    public int getPriceMonthly() { return priceMonthly; }

    @Override
    public String toString() {
        return "PlanLimits{" +
                "maxWorkflows=" + maxWorkflows +
                ", maxModels=" + maxModels +
                ", maxExecutionsPerMonth=" + maxExecutionsPerMonth +
                ", maxTeamMembers=" + maxTeamMembers +
                ", reqPerMinute=" + reqPerMinute +
                ", burstCapacity=" + burstCapacity +
                ", maxWebhooksPerDay=" + maxWebhooksPerDay +
                ", priceMonthly=" + priceMonthly +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        PlanLimits that = (PlanLimits) o;
        return maxWorkflows == that.maxWorkflows &&
                maxModels == that.maxModels &&
                maxExecutionsPerMonth == that.maxExecutionsPerMonth &&
                maxTeamMembers == that.maxTeamMembers &&
                reqPerMinute == that.reqPerMinute &&
                burstCapacity == that.burstCapacity &&
                maxWebhooksPerDay == that.maxWebhooksPerDay &&
                priceMonthly == that.priceMonthly;
    }

    @Override
    public int hashCode() {
        return Objects.hash(maxWorkflows, maxModels, maxExecutionsPerMonth,
                maxTeamMembers, reqPerMinute, burstCapacity, maxWebhooksPerDay, priceMonthly);
    }
}

File: flowforge-ui/src/config/planLimits.ts
/**
 * Central plan limits configuration — single source of truth for the frontend.
 * Must stay in sync with PlanLimits.java in flowforge-common.
 *
 * A value of -1 means "unlimited".
 */

export type Plan = 'FREE' | 'PRO' | 'ENTERPRISE'

export interface PlanLimitsConfig {
  maxWorkflows: number
  maxModels: number
  maxExecutionsPerMonth: number
  maxTeamMembers: number
  reqPerMinute: number
  burstCapacity: number
  maxWebhooksPerDay: number
  priceMonthly: number // cents (0 = free, -1 = custom)
}

export const PLAN_LIMITS: Record<Plan, PlanLimitsConfig> = {
  FREE: {
    maxWorkflows: 3,
    maxModels: 5,
    maxExecutionsPerMonth: 1_000,
    maxTeamMembers: 2,
    reqPerMinute: 10,
    burstCapacity: 20,
    maxWebhooksPerDay: 100,
    priceMonthly: 0,
  },
  PRO: {
    maxWorkflows: 25,
    maxModels: 50,
    maxExecutionsPerMonth: 100_000,
    maxTeamMembers: 10,
    reqPerMinute: 60,
    burstCapacity: 100,
    maxWebhooksPerDay: 10_000,
    priceMonthly: 4_900,
  },
  ENTERPRISE: {
    maxWorkflows: -1,
    maxModels: -1,
    maxExecutionsPerMonth: -1,
    maxTeamMembers: -1,
    reqPerMinute: 600,
    burstCapacity: 1_000,
    maxWebhooksPerDay: -1,
    priceMonthly: -1,
  },
}

/** Human-readable plan metadata for UI cards, comparison grids, etc. */
export const PLAN_META: Record<Plan, { label: string; price: string; tagline: string }> = {
  FREE: { label: 'Free', price: '$0/mo', tagline: 'For personal projects and exploration' },
  PRO: { label: 'Pro', price: '$49/mo', tagline: 'For growing teams and production workloads' },
  ENTERPRISE: { label: 'Enterprise', price: 'Custom', tagline: 'Dedicated infra, SLA guarantee, priority support' },
}

/** Format a limit value for display — returns "Unlimited" for -1 */
export function formatLimit(value: number): string {
  if (value === -1) return 'Unlimited'
  return value.toLocaleString()
}

/** Check if a usage count has hit the limit (false when unlimited) */
export function isAtLimit(used: number, limit: number): boolean {
  return limit !== -1 && used >= limit
}

/** Check if usage is near the limit (>= 80%) */
export function isNearLimit(used: number, limit: number): boolean {
  if (limit === -1) return false
  return used >= limit * 0.8
}

/** Get the list of feature rows for plan comparison grids */
export function getPlanFeatureRows(): Array<{ label: string; key: keyof PlanLimitsConfig; unit?: string }> {
  return [
    { label: 'Workflows', key: 'maxWorkflows' },
    { label: 'Data Models', key: 'maxModels' },
    { label: 'Executions / month', key: 'maxExecutionsPerMonth' },
    { label: 'Team Members', key: 'maxTeamMembers' },
    { label: 'API Rate Limit', key: 'reqPerMinute', unit: 'req/min' },
    { label: 'Webhooks / day', key: 'maxWebhooksPerDay' },
  ]
}
